cli/runner.go

```diff
@@ -0,0 +1,305 @@
+	// Internal bool     `json:"internal"`
+	// DNS      bool     `json:"dns_enabled"`
+	// IpamOpt  IpamOpt  `json:"ipam_options"`
+}
+
+type DockerNetwork struct {
+	Name   string `json:"Name"`
+	Id     string `json:"Id"`
+	Driver string `json:"Driver"`
+	// 	Scope    string   `json:"Scope"`
+	// 	Created  string   `json:"Created"`
+	// 	IPv6     bool     `json:"EnableIPv6"`
+	// 	Internal bool     `json:"Internal"`
+	// 	Attachable
+	// 	Ingress
+	// 	Containers
+	// 	DNS      bool     `json:"dns_enabled"`
+	Ipam Ipam `json:"IPAM"`
+}
+
+type Ipam struct {
+	//Driver
+	//Options
+	Config []Subnet
+}
+
+type Subnet struct {
+	Subnet  string `json:"subnet"`
+	Gateway string `json:"gateway"`
+}
+type IpamOpt struct {
+	Driver string `json:"driver"`
+}
+
+func GetCNIGateway() (error, string) {
+	//get networks
+
+	err, executable := findExec(false)
+	if err != nil {
+		return err, ""
+	}
+
+	namefilter := fmt.Sprintf("name=%s", ContainerNetName)
+
+	cmd := exec.Command(executable, "network", "ls", "-f", "driver=bridge", "-f", namefilter, "--format", "'{{.Name}}'")
+
+	cmdOutput := &bytes.Buffer{}
+	cmd.Stdout = cmdOutput
+	cmd.Stderr = os.Stderr
+
+	log.Println("Executing", executable, "in the background, listing network...")
+	err = cmd.Run()
+	if err != nil {
+		return err, ""
+	}
+
+	if cmdOutput.String() == "" {
+		err = EnsureDyoNetwork(executable, true)
+	}
+
+	return GetGatewayIP(executable)
+}
+
+func GetGatewayIP(executable string) (error, string) {
+
+	// inspect, get gw addr
+	cmd := exec.Command(executable, "network", "inspect", ContainerNetName)
+
+	cmdOutput := &bytes.Buffer{}
+	cmd.Stdout = cmdOutput
+	cmd.Stderr = os.Stderr
+
+	log.Println("Executing", executable, "in the background, inspecting network...")
+	err := cmd.Run()
+	if err != nil {
+		return err, ""
+	}
+
+	switch executable {
+	case "podman.exe":
+		fallthrough
+	case "podman":
+		var network []PodmanNetwork
+		err = json.Unmarshal(cmdOutput.Bytes(), &network)
+		if err != nil {
+			return err, ""
+		}
+
+		if len(network) != 1 {
+			return errors.New("there are more than one network"), ""
+		}
+
+		if len(network[0].Subnets) != 1 {
+			return errors.New("there are more than one network subnets"), ""
+		}
+
+		gwip := network[0].Subnets[0].Gateway
+
+		if gwip != "" {
+			return nil, gwip
+		}
+	case "docker.exe":
+		fallthrough
+	case "docker":
+		var network []DockerNetwork
+		err = json.Unmarshal(cmdOutput.Bytes(), &network)
+		if err != nil {
+			return err, ""
+		}
+
+		if len(network) != 1 {
+			return errors.New("there are more than one network"), ""
+		}
+
+		if len(network[0].Ipam.Config) != 1 {
+			return errors.New("there are more than one network subnets"), ""
+		}
+
+		gwip := network[0].Ipam.Config[0].Gateway
+
+		if gwip != "" {
+			return nil, gwip
+		}
+	default:
+		return errors.New("Unknown binary"), ""
+	}
+
+	return errors.New("gateway IP is empty"), ""
+}
+
+func EnsureDyoNetwork(executable string, delnet bool) error {
```

Comment:
I am just curious what is delnet?

Reply:
Network deletion if it exists with wrong parameters like wrong driver. Now it defaults to true, planned to be changeable.